### DISCUSSION TITLE
Pre-installed vs. On-demand System Packages

### DISCUSSION THREAD

**Discussion by colincasey** [heroku/buildpacks#8] [Ideas]
Our [Heroku base images](https://github.com/heroku/base-images) come with a [preinstalled set of packages](https://devcenter.heroku.com/articles/stack-packages). This list of packages attempts to include build and runtime dependencies frequently needed by applications but it can't include every possible package. 

For applications that require packages not included in the [Heroku base images](https://github.com/heroku/base-images), we plan to create a Cloud Native Buildpack similar to the [Heroku Apt Buildpack](https://github.com/heroku/heroku-buildpack-apt). This would allow applications to install packages on demand. 

The main difference between the pre-installed versus on-demand packages is:
- Pre-installed packages are automatically kept up to date with [regular base image updates](https://devcenter.heroku.com/articles/stack-update-policy).
- On-demand packages are only updated when an application is rebuilt.

Please use this discussion to provide thoughts, links to tools, or feedback on the following topics:
- Are there packages that should be added to our pre-installed set?
- Are there packages that should be be removed from our pre-installed set?
- Should we allow packages to be installed from non-standard Debian repositories?
- Should we allow packages to be installed from Debian urls?

**Comment by schneems** [comment]
In https://github.com/heroku/buildpacks/discussions/6#discussioncomment-8833294 it would be good to have an easy/documented way to install jemalloc. It's a popular memory allocator that works well with Ruby applications. I see there is a `libjemalloc2` ubuntu package. I'm not sure if that's all that's needed. There is a [classic jemalloc buildpack](https://github.com/gaffneyc/heroku-buildpack-jemalloc) that also populates a script for setting `LD_PRELOAD`as well:

```
~ $ cat vendor/jemalloc/bin/jemalloc.sh
#!/bin/sh

prefix=/app/vendor/jemalloc
exec_prefix=/app/vendor/jemalloc
libdir=${exec_prefix}/lib

LD_PRELOAD=${libdir}/libjemalloc.so.2
export LD_PRELOAD
exec "$@"
```
